src/services/diff.ts

```diff
@@ -1,31 +1,43 @@
+function mapSegmentation<X, Y>(a: Map<X, Y>, b: Map<X, Y>): [Map<X, Y>, Map<X, [Y, Y]>, Map<X, Y>] {
+  const left: Map<X, Y> = new Map();
+  const intersect: Map<X, [Y, Y]> = new Map();
+  const right: Map<X, Y> = new Map();
+  // I feel like this can be done with a single pass, but I'm not sure exactly how, yet. At least it's
+  // not three passes
+  for (const aVal of a) {
+    if (b.has(aVal[0])) intersect.set(aVal[0], [aVal[1], b.get(aVal[0])!]);
+    else left.set(...aVal);
+  }
+  for (const bVal of b) {
+    if (!a.has(bVal[0])) right.set(...bVal);
+  }
+  return [left, intersect, right];
```

Comment:
is there a way enforce or passively keep track of the timings for this module to make sure there are not regressions on this logic?